### PR TITLE
Session Specific Transient Directory

### PR DESCRIPTION
**Pull request by russellpierce**
An attempt to address https://github.com/robertzk/lockbox/issues/8.  Passes local runs of the tests with `package:uuid` installed and without `package:uuid` installed.  I haven't added any tests specific to this modification yet, but I am open to suggestions.

### CODE REVIEW

File: R/zzz.R
Comment (robertzk) [inline R/zzz.R]: This only errors if it doesn't exist, right? What about

```r
if (exists(target, envir = .lockbox_env, inherits = FALSE)) {
  get(target, envir = .lockbox_env)
} else {
  default
}
```

If that is the intended behavior, I suggest `.lockbox_env$session_id %||% default` instead of the function, which should be equivalent. Because of my mistake in defining `.lockbox_env`, your current approach could run into the snag where `target` is define in a parent environment: if there is a global variable `session_id` that will be used instead! Maybe you can fix zzz.R for me:

```r
.lockbox_env <- new.env(parent = emptyenv())
```

File: R/zzz.R
Comment (robertzk) [inline R/zzz.R]: Does `.lockbox_env$session_id <- .lockbox_env$session_id %||% uuid::UUIDgenerate()` operate differently? I was thinking these may be the same while also avoiding the outside `if` null check. It would also be consistent with other

File: R/zzz.R
Comment (robertzk) [inline R/zzz.R]: I would prefix with another `#` which will get placed on the left hand side of the [rocco documentation](http://robertzk.github.io/lockbox/).

File: R/lockbox.R
Comment (robertzk) [inline R/lockbox.R]: We probably want `~/.R/lockbox_transient/<session_id>` here with something like `default = "global"` below.

File: R/zzz.R
Comment (robertzk) [inline R/zzz.R:104]: 👍

File: R/zzz.R
Comment (robertzk) [inline R/zzz.R]: I like this solution.

Maybe use the helpers here to avoid problems if the paths change:

```r
# Note the paths should already be normalized here.
lapply(c(lockbox_transient_staging_dir(), lockbox_transient_dir(), lockbox_download_dir()), 
  function(path) {  try(unlink(x, recursive = TRUE, force = TRUE), silent = TRUE) })
```

I would even define a helper `lockbox_session_dirs` in [lockbox.R](https://github.com/robertzk/lockbox/blob/cf10b065e770e1cc5066b730985f40d494d3bcc7/R/lockbox.R#L145) for the above 3-vector.

File: .travis.yml
Comment (robertzk) [inline .travis.yml]: Can you run `git checkout origin .travis.yml` to restore this?

**Review by robertzk — CHANGES_REQUESTED** [review]
Looks solid decision-wise but haggling over a few minor style comments.

**Comment by robertzk** [comment]
@drknexus Looks like travis is running after the PR was opened (or that might just be Zapier's...). Let's wait until it passes and I will merge this.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: You're right.  I don't know why I got in the pattern of get() for a target environment rather than using `$`... which kindly returns a NULL anyway rather than throwing an error.

Comment (robertzk) [reply R/zzz.R]: Agree that the multiple equivalents are confusing. I typically only find myself using `get` if I need to specify `inherits = FALSE`.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: I want to be sure we don't risk getting a new session_id in the middle of a session.  I'm thinking of adding an early escape `if (!is.null(.lockbox_env$session_id)) {return()}` like you've done elsewhere to help be sure of that, then for the rest of the code we aren't stuck a level of nesting deep.  This also gets the symlink cleanup inside the protection which is where it belongs.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: This is my first time working with the finalizers.  Based on my reading so far, it is getting called because something tried to destroy `.lockbox_env`.  I was/am not certain how much of `package:lockbox` will be callable at that point.  I'd have to do that testing tomorrow.

File: .travis.yml
Comment (russellpierce) [reply .travis.yml]: Will do.

**Comment by russellpierce** [comment]
It looks like I got confused about Travis based on a bad experience previously.  It runs on your Travis when I open the PR - probably using the Travis settings from the PR.  So long as you don't mind those notifications, things should be fine.  It looks like I can keep Travis from running on my fork - it just means that CI is deferred until I open the PR. :/

File: R/zzz.R
Comment (robertzk) [reply R/zzz.R]: Use of the [`%||%` infix](https://github.com/robertzk/lockbox/blob/master/R/utils.R#L1) should cover that, right? It checks implicitly for whether the left-hand side is null, and if it is, only then proceeds to use the right hand side.

File: R/zzz.R
Comment (robertzk) [reply R/zzz.R]: Yeah this is a good question. I just tested [unload hooks](https://stat.ethz.ch/R-manual/R-devel/library/base/html/ns-hooks.html) and it doesn't look like they run if you `q()` directly. I'll do some testing to ensure this covers most sane cases of the R interpreter exiting.

File: R/lockbox.R
Comment (russellpierce) [reply R/lockbox.R]: I was having trouble reading the intent on `lockbox_transient`.  Does the useR expect the `lockbox.transient_dir` to point unadulterated to their transient directory or do they think it is the root where we will dump sessions?  I was thinking the former.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: Yes, I think you're right that it will cover the assignment to `.lockbox$session_id`.  I do think a bit of something to make sure we only register the finalizer a single time is called for though.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: Is it alright to leave the Rocco build to you?  I don't have experience with that package yet.

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: I think I've addressed these comments with 3c78e19 and bcd07cb

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: This bit is currently a bit of a mess because I'm not quite sure where to stick the finalizer / how we'll finalize is a bit in flux.

File: .travis.yml
Comment (russellpierce) [reply .travis.yml]: Should be addressed with 633ffd2344c2dd712cdd05cae71230a05f57bbd1

File: R/zzz.R
Comment (abelcastilloavant) [inline R/zzz.R]: Don't know what the message means - maybe it's incomplete?

**Comment by russellpierce** [comment]
Of the issues above, I think the only major ones left were:

1. You want the session transients to be under ~/.R/lockbox_transient/{session_id}.  To move forward I need clarification about whether when `lockbox.transient_dir` is set whether you want session_id to be respected and added as a subdirectory or neglected.
2.  Cleaning up the session specific directories regardless of their name.  I'll test if the lockbox functions are available later today.  If they aren't, I was considering storing them in .lockbox_env which will certainly be available at the finalizer since that is the object about to be destructed.

... did I miss any that you see as blocking?

**Review by abelcastilloavant — COMMENTED** [review]
looks good to me. Thanks @drknexus!

File: R/zzz.R
Comment (russellpierce) [reply R/zzz.R]: @abelcastilloavant Thanks for catching that.  It was an incomplete thought.  I'd decided not to send a message there.  I'll drop that message.

**Comment by russellpierce** [comment]
Assuming the answer for #1 was 'neglected', for #2 I found that lockbox functions were still available  🤷‍♂️ .  As far as I know that clears all outstanding issues.

**Comment by robertzk** [comment]
@drknexus Apologies for review delays. Been slammed with end of Q2. 

1. I think `lockbox.transient_dir` should be respected exactly as given by the user to avoid surprises, since it can be used to perform "manual sessions" if the user isn't happy with the existing approach. I agree this sets them up for a potential trap if they use the same transient directory in two competing custom sessions, but I feel the onus is on the user there.

2. Do you mean the package namespace functions? I don't think the search path is [`unloaded`](https://stat.ethz.ch/R-manual/R-devel/library/base/html/ns-hooks.html) before the R session exists, so they should be available. 

Otherwise it looks good. I'll merge it in and if I find any issues we'll address that in a separate PR. Thanks for the all work!

**Comment by robertzk** [comment]
Ran it on a fresh install with two big lockfiles / Syberia projects and keep running into some problems downloading like below but it seems to be sporadic. I'll [tag it](https://github.com/robertzk/lockbox/tree/0.2.4.5) for now with the version on my remote so you can start using it, and try one more time tomorrow before merging to confirm it's just network issues. 

```
Error in read.dcf(file.path(pkg_path, "DESCRIPTION")) :
  cannot open the connection
In addition: Warning message:
running command '/usr/bin/gzip -dc '/Users/robertk/.R/lockbox/lockbox_download_dir/gbm/CRAN2.1.1.tar.gz
' | /usr/bin/tar -tf '-'' had status 1

```

**Comment by russellpierce** [comment]
The download dir defaults to ~.R/lockbox/lockbox_download_dir, that is, not under the transient tree. So, multiple sessions can collide on download in the current revision. The assumption is that the lockbox comes prefilled. But, arguably it would be nice/good to have session based download directories too. I can look into tweaking this further or we can do that as a new issue.

**Comment by robertzk** [comment]
Good point. The download directory should definitely be versioned too. If you can add the session ID to the download dir in this PR, I'll test it in the morning. Thanks!

**Comment by russellpierce** [comment]
The reason I didn't go all of the way before is I didn't take the time sit down and think through what happens on the downstream edges... e.g., what happens when multiple sessions trigger `move_package_to_lockbox_library()` at the same time?  

A quick glance suggests that `unlink()` doesn't mind if its target is non-existent... `file.rename()` throws a warning if the source is missing and returns FALSE.  As far as I can tell nothing in the code will mind much - just a mysterious warning might leak out.

But, that's a real quick check on my part.  I still wouldn't want to promise that simultaneous downloads is safe/robust, but this change should make them perhaps a little less likely to cause problems than they were before.  I guess we'll know more with your tests in the morning. :)

**Comment by robertzk** [comment]
@drknexus Agreed. I'll be out for a July 4th camping trip but will do some more thorough thinking after the break.

**Pull request merged**
